### PR TITLE
Revert "nixos: default `environment.homeBinInPath` to false"

### PR DESCRIPTION
**Pull request by edolstra**
This reverts commit a06529b7adbfe39e06c10d0c539418d130f2fecc. Having `~/bin` in `$PATH` has been the case in NixOS since forever or so, so I don't see a strong reason for changing this. People who don't want this can set `environment.homeBinInPath` to false now.

### CODE REVIEW

**Review by jonringer — APPROVED** [review]
would still like @andir opinion

**Comment by andir** [comment]
I would argue that the more common case is to NOT have it on your path and if you need that you can opt-in to it. We have a proper release note entry documenting it. If we revert this how can we ever deprecate something?

**Comment by rnhmjoj** [comment]
I don't really see the point of this option being in NixOS. It's something that can be done trivially in the user `.profile` or via `environment.variables.PATH` in NixOS.

**Comment by edolstra** [comment]
We've had this discussion before and the result was not to remove it (https://github.com/NixOS/nixpkgs/issues/40065). IMHO nothing has really changed since then.

**Comment by dasJ** [comment]
-1 on this

Encouraging impurities like that should be kept to a minimum by default

**Comment by andir** [comment]
Looking at the old discussion we did lack a proper way of dealing with deprecation. I think we do have one and that is called release notes and a way to revert to the old (potentially never desired behavior).  The change is listed in the category for breaking changes and has an almost copy-able line that enables the old mode of operation. (`If you depend on this you should set the option <literal>environment.homeBinInPath</literal> to <literal>true</literal>.`)

We have had two users (you and @xeji) of use this. I am not sure if @xeji was aware of the feature since he stated it comes from his dotfiles repo anyway.

The majority of people reacting to see seem to be against this PR - as far as you can trust GH reactions…
I am missing really good arguments why this is a good default for everyone. Ignoring the fact that it was default for many years now.

Once upon a time I also had a `~/bin` folder but ever since I switched to NixOS I like my scripts to be not mutable anymore… Not sure if that counts as a counter argument.

**Comment by adisbladis** [comment]
`environment.homeBinInPath = true` is a _highly_ surprising default, especially given that we strive for purity.

I'm strongly against this PR.

**Comment by jonringer** [comment]
> `environment.homeBinInPath = true` is a _highly_ surprising default, especially given that we strive for purity.

On the flip side, people trying nixos might find it highly surprising that it's not there. I think it's just a matter of perspective. For the purity argument, I agree that the NixOS system probably shouldn't have any knowledge of a given user's HOME directory.

I guess I don't feel strongly about this. Those that need `~/bin` on their path will probably do the correct steps to ensure it's there, while those don't use it will likely never know it's there.

**Comment by joepie91** [comment]
> On the flip side, people trying nixos might find it highly surprising that it's not there.

On the other hand, has the ship of non-surprising behaviour not already kinda sailed considering there's effectively no `/bin` or `/lib` either?

I mean, we have basically two options here:
1. Add another "difference with other distros" to the list, but a difference that is consistent with the NixOS design rationale.
2. Avoid differing from other distros for this one small bit, and then have to explain why this is an exception to the overall design rationale.

I feel like considering the circumstances, option 1 would probably be preferable?

**Comment by gilligan** [comment]
Opting for `~/bin` to be in PATH simply because it has always been that way isn't very convincing to me. Why isn't _now_ the right time to do the right thing in this regard? There aren't currently any rules regarding a stable API or anything like that which would prevent this change (?).

**TL;DR**: I'd prefer for this to *not* be merged. Not having `~/bin` in PATH sounds more desirable to me.

**Comment by jonringer** [comment]
Related to the "similar to other distro's" topic is a discussion that already occurred on discourse: https://discourse.nixos.org/t/why-i-switched-to-manjaro-linux/5903/22.

If we decide to include such a thing in nixos, people who want a more "batteries-included" experience can opt into importing a "profile", which will create a desktop environment with all the defaults take from opinionated distro. 

E.g:
```
  imports = [
    ./hardware-configuration.nix
    <nixos/profiles/ubuntu>
  ];
```

**Comment by edolstra** [comment]
The issue is not what other Linux distros do, it's what NixOS has done for the last decade or so. If you don't want ~/bin to be used, the simple solution is to not have a ~/bin.

**Comment by edolstra** [comment]
I'll bow to the will of the people and close this PR.
